### PR TITLE
fix(sql): fix SAMPLE BY queries with functions in SELECT clause

### LINKED ISSUE #4315: Sample by and to_timezone
### To reproduce

The to_timezone function is not liked by sample by. On demo 

```SQL
select symbol,sum(amount), vwap(price, amount), to_timezone(timestamp,'EST') NYTime
from trades
where timestamp in '2024-03-19' 
sample by 5m;
```

returns 2m rows while

```SQL
select symbol,sum(amount),vwap(price, amount), timestamp
from trades
where timestamp in '2024-03-19' 
sample by 5m;
```

returns 7k rows. 
 

gives 592,502 records , ignores the sample by

### QuestDB version:

7.3.10

### OS, in case of Docker specify Docker and the Host OS:

Linux

### File System, in case of Docker specify Host File System:

ext4

### Full Name:

Alex Pelagenko

### Affiliation:

QuestDB

### Have you followed Linux, MacOs kernel configuration steps to increase Maximum open files and Maximum virtual memory areas limit?

- [X] Yes, I have

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by ideoma**
Fixes #4315

Fixes SQL where functions are used with timestamp column argument in SAMPLE BY queries, for example 

```SQL
select symbol || 'abcd' as symbol, sum(amount), vwap(price, amount), 
dateadd('h', 2, to_timezone(timestamp,'EST')) + 1000000L NYTime
from trades
sample by 5m
```

Note, there is a TODO left to fix query:

```SQL
select symbol || 'abcd' as symbol, sum(amount), vwap(price, amount),
cast(dateadd('h', 2, to_timezone(timestamp,'EST')) as double) as ts,
cast(dateadd('h', 2, to_timezone(timestamp,'EST')) as double) + sum(amount) NYTime
from trades
sample by 5m
```

### CODE REVIEW

**Comment by ideoma** [comment]
#### [PR Coverage check]

:heart_eyes: **pass** : 81 / 81 (100.00%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/griffin/SqlOptimiser.java|81|81|100.00%|

File: core/src/main/java/io/questdb/griffin/SqlOptimiser.java [hunk 1]
```diff
@@ -2721,6 +2734,47 @@ private ExpressionNode nextLiteral(CharSequence token) {
         return nextLiteral(token, 0);
     }
 
+    private boolean nonAggregateFunctionDependsOn(ExpressionNode node, CharSequence argument) {
+        sqlNodeStack.clear();
+        while (node != null) {
+            if (node.type != FUNCTION) {
```
Comment (jerrinot) [inline core/src/main/java/io/questdb/griffin/SqlOptimiser.java:2740]: nit: `node.type != FUNCTION` appears to be redundant?

File: core/src/main/java/io/questdb/griffin/SqlOptimiser.java [hunk 2]
```diff
@@ -2721,6 +2734,47 @@ private ExpressionNode nextLiteral(CharSequence token) {
...
+                        sqlNodeStack.push(node.lhs);
+                    }
+
+                    if (node.rhs != null) {
+                        node = node.rhs;
+                        continue;
+                    }
+                } else {
+                    for (int i = 1, k = node.paramCount; i < k; i++) {
+                        if (node.args.getQuick(i) != null) {
```
Comment (jerrinot) [inline core/src/main/java/io/questdb/griffin/SqlOptimiser.java:2758]: more for my enlightenment: can this ever be `null`? and if so, can `node.args.getQuick(0)` bellow be `null` too? if so, is it possible we exit this method with a non-empty stack and thus reporting possibly false-negative?

Comment (ideoma) [reply core/src/main/java/io/questdb/griffin/SqlOptimiser.java:2758]: It shouldn't be null, we can remove the if.

**Review by jerrinot — DISMISSED** [review]

**Review by bluestreak01 — DISMISSED** [review]

**Comment by ideoma** [comment]
#### [PR Coverage check]

:heart_eyes: **pass** : 79 / 79 (100.00%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/griffin/SqlOptimiser.java|79|79|100.00%|

**Review by jerrinot — APPROVED** [review]

**Comment by ideoma** [comment]
#### [PR Coverage check]

:heart_eyes: **pass** : 79 / 79 (100.00%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/griffin/SqlOptimiser.java|79|79|100.00%|

**Pull request merged**